# Case Study - Group 11 (S11)

### Group Members
- **Darylle Xander Calugtong**
- **Graham Joshua Ogatia**
- **Joshua Permito**
- **Shem Matthew Salih**

## Dataset Information
Our chosen dataset is the [Valorant Leaderboard Statistics](https://www.kaggle.com/datasets/aliibrahim10/valorant-stats).
This dataset contains [tracker.gg](https://tracker.gg/) statistics with 80,000+ observations and 38 variables
from **Episode 4: Act 3** (April 2022) of the game **Valorant**.
### Important
The observations in the dataset only contain players who are **Immortal** and **Radiant** rank except for a few players
who were demoted at the time of data collection.
Additionally, only those who consent for the platform [tracker.gg](https://tracker.gg) to gather their statistics have their data in the dataset.
### Collection
The platform [tracker.gg](https://tracker.gg/) collects the data from the game **Valorant**.
This site is a third-party platform that tracks the statistics of players in the game.
It is a third-party platform that is not affiliated with the game developers.

## Dataset Structure
Each row represents a player's statistics in the game **Valorant**.
Each column represents a variable that describes the player's statistics.

### Variables
- `region` - The region of the player.
- `name` - The name of the player. (used together with `tag`)
- `tag` - The tag of the player. (used together with `name`)
- `rating` - The rank of the player.
- `damage_round` - The average damage per round of the player in all matches.
- `headshots` - Total number of headshots of the player in all matches.
- `headshot_percent` - Percentage of total hit shots that are headshots of the player in all matches.
- `aces` - Total number of aces of the player in all matches. An ace is when a player kills all 5 players in the enemy team in a single round.
- `clutches` - Total number of clutches of the player in all matches. A clutch is when a player wins a round when they are the last player alive in their team.
- `flawless` - Total number of flawless rounds of the player in all matches. A flawless round is when a team wins a round without any of their players dying.
- `first_bloods` - Total number of first bloods of the player in all matches.
- `kills` - Total number of kills of the player in all matches.
- `deaths` - Total number of deaths of the player in all matches.
- `assists` - Total number of assists of the player in all matches.
- `kd_ratio` - The kill-death ratio of the player in all matches.
- `kills_round` - Average number of kills per round of the player in all matches.
- `most_kills` - Most number of kills in a single match of the player in all matches.
- `score_round` - Average score per round of the player in all matches.
- `wins` - Total number of wins of the player in all matches.
- `win_percent` - Percentage of total matches won of the player in all matches.
- `agent_1` - Most played agent of the player in all matches based on hours.
- `agent_2` - Second most played agent of the player in all matches based on hours.
- `agent_3` - Third most played agent of the player in all matches based on hours.
- `gun1_name` - Player's best gun in all matches based on kills.
- `gun1_head` - Percentage of total hit shots that are headshots of the player using `gun1_name` in all matches.
- `gun1_body` - Percentage of total hit shots that are body shots of the player using `gun1_name` in all matches.
- `gun1_legs` - Percentage of total hit shots that are leg shots of the player using `gun1_name` in all matches.
- `gun1_kills` - Total number of kills of the player using `gun1_name` in all matches.
- `gun2_name` - Player's second-best gun in all matches based on kills.
- `gun2_head` - Percentage of total hit shots that are headshots of the player using `gun2_name` in all matches.
- `gun2_body` - Percentage of total hit shots that are body shots of the player using `gun2_name` in all matches.
- `gun2_legs` - Percentage of total hit shots that are leg shots of the player using `gun2_name` in all matches.
- `gun2_kills` - Total number of kills of the player using `gun2_name` in all matches.
- `gun3_name` - Player's third-best gun in all matches based on kills.
- `gun3_head` - Percentage of total hit shots that are headshots of the player using `gun3_name` in all matches.
- `gun3_body` - Percentage of total hit shots that are body shots of the player using `gun3_name` in all matches.
- `gun3_legs` - Percentage of total hit shots that are leg shots of the player using `gun3_name` in all matches.
- `gun3_kills` - Total number of kills of the player using `gun3_name` in all matches.

## Preprocessing

### Import Libraries
First, we need to import the necessary libraries.

We will be using the following libraries:
- `pandas` - for data manipulation
- `numpy` - for mathematical operations
- `mplcatppuccin` - matplotlib theme (installed as `catppuccin-jupyterlab`)
- `matplotlib` - for data visualization

In [1]:
import pandas as pd
import numpy as np
import mplcatppuccin
import matplotlib.style as mplstyle
import matplotlib.pyplot as plt

Set the theme to `mocha` from catppuccin

In [2]:
mplstyle.use('mocha')

### Importing the Dataset

In [3]:
# Only consider blank cells as NA, and thousands to specify the thousands separator
val_df = pd.read_csv('val_stats.csv', thousands=',', keep_default_na=False, na_values=[''],
                     dtype={'region':'category', 'rating':'category'})
val_df

region            name     tag      rating  damage_round  headshots  \
0         NA        ShimmyXD    #NA1     Radiant         135.8        992   
1         NA       XSET Cryo  #cells     Radiant         170.3        879   
2         NA   PuRelittleone  #yoruW     Radiant         147.5        720   
3         NA            Boba   #0068     Radiant         178.2        856   
4         NA     i love mina  #kelly     Radiant         149.8        534   
...      ...             ...     ...         ...           ...        ...   
85673    LAT        Kazutora  #img0d     Radiant         138.2        342   
85674    LAT  el lobo marino  #uthur     Radiant         182.9        650   
85675    LAT          p9pzet   #666x     Radiant         158.8        613   
85676    LAT  EZ4TGD EnSBuwu  #kmeve  Immortal 3         155.9        132   
85677    LAT            Neon    #SSJ     Radiant         164.4       1127   

       headshot_percent  aces  clutches  flawless  ...  gun2_name  gun2_head  \
0                  24.9     0       140        80  ...    Phantom         33   
1                  28.3     2       122        94  ...   Operator          8   
2                  24.0     3       117        59  ...    Phantom         36   
3                  37.3     3        83        49  ...    Sheriff         48   
4                  24.4     2        71        38  ...    Spectre         21   
...                 ...   ...       ...       ...  ...        ...        ...   
85673              21.4     0        58        54  ...     Vandal         28   
85674              30.1     4        77        42  ...     Vandal         40   
85675              30.2     0        70        54  ...    Phantom         40   
85676              22.2     0        23        18  ...    Phantom         39   
85677              20.7     3       136        82  ...   Operator          9   

       gun2_body  gun2_legs  gun2_kills  gun3_name  gun3_head  gun3_body  \
0             62          5         220    Classic         36         60   
1             91          0         226    Phantom         32         63   
2             61          3         231   Operator          8         91   
3             51          1          48    Phantom         44         56   
4             71          8          65   Operator          8         92   
...          ...        ...         ...        ...        ...        ...   
85673         69          2         175    Classic         39         59   
85674         57          3         212    Spectre         33         63   
85675         56          4         159   Operator         10         89   
85676         57          3          37    Spectre         27         69   
85677         88          2         132    Spectre         27         70   

       gun3_legs  gun3_kills  
0              3         147  
1              5         137  
2              1         102  
3              0          36  
4              0          64  
...          ...         ...  
85673          2          71  
85674          5         139  
85675          1          87  
85676          4          17  
85677          2         108  

[85678 rows x 38 columns]

### List of all variables

In [4]:
[col for col in val_df.columns]

['region',
 'name',
 'tag',
 'rating',
 'damage_round',
 'headshots',
 'headshot_percent',
 'aces',
 'clutches',
 'flawless',
 'first_bloods',
 'kills',
 'deaths',
 'assists',
 'kd_ratio',
 'kills_round',
 'most_kills',
 'score_round',
 'wins',
 'win_percent',
 'agent_1',
 'agent_2',
 'agent_3',
 'gun1_name',
 'gun1_head',
 'gun1_body',
 'gun1_legs',
 'gun1_kills',
 'gun2_name',
 'gun2_head',
 'gun2_body',
 'gun2_legs',
 'gun2_kills',
 'gun3_name',
 'gun3_head',
 'gun3_body',
 'gun3_legs',
 'gun3_kills']

## Data Cleaning

### Check for Multiple Representations
Let's proceed to check for multiple representations of values. We will be focusing on categorical variables of `rating`, `region`, `agent_1`, and `gun1_name`.

For the `rating`, let's first list down all the unique values of the variable.

In [35]:
ratings_df = pd.DataFrame(val_df['rating'].unique(), columns=['Rating'])
ratings_df

Rating
0      Radiant
1   Immortal 3
2   Immortal 2
3   Immortal 1
4     Silver 2
5       Gold 2
6    Diamond 3
7       Gold 1
8      Unrated
9   Platinum 1
10    Bronze 3
11      Gold 3
12    Bronze 2
13   Diamond 2
14    Silver 3
15    Silver 1
16  Platinum 2
17   Diamond 1
18  Platinum 3

Valorant follows the [ranking system](https://support-valorant.riotgames.com/hc/en-us/articles/360047937633-VALORANT-Competitive-Mode-FAQ) of **Iron**, **Bronze**, **Silver**, **Gold**, **Platinum**, **Diamond**, **Ascendant**, **Immortal**, and **Radiant**. Excluding **Radiant**, each rank is divided into three tiers of **1**, **2**, and **3** wherein the higher the number reflects better ranking. Players who do not have a rank are considered **Unrated**. With that in mind, the values of `rating` are considered valid as they are all within the game's ranking system. It could also be observed that there no multiple representations of a rank. 

Now for `region`, `agent_1`, and `gun1_name`, let's check for their representation and validity.

In [6]:
regions_df = pd.DataFrame(val_df['region'].unique(), columns=['Region'])
regions_df

Region
0     NA
1     EU
2     AP
3     KR
4     BR
5    LAT

In [29]:
agent_1_df = pd.DataFrame(val_df['agent_1'].unique(), columns=['Agent'])
agent_1_df

Agent
0        Fade
1     Chamber
2        Yoru
3        Jett
4        Sage
5       KAY/O
6        Sova
7        Raze
8        Omen
9      Breach
10      Reyna
11       Neon
12       Skye
13      Viper
14  Brimstone
15    Phoenix
16      Astra
17    Killjoy
18     Cypher

In [8]:
gun1_name_df = pd.DataFrame(val_df['gun1_name'].unique(), columns=['Gun'])
gun1_name_df

Gun
0     Vandal
1   Operator
2    Phantom
3      Judge
4       Odin
5   Guardian
6    Spectre
7    Classic
8      Ghost
9    Sheriff
10   Marshal
11    Shorty
12   Bulldog
13      Ares
14    Frenzy
15     Bucky

From the results above, we could observe 6 `region`, 19 `agent_1`, and 16 `gun1_name` values with no multiple representation. It is also important to note that the `region` values listed are considered valid [servers](https://support-valorant.riotgames.com/hc/en-us/articles/360055678634-Server-Select) in the game (NA - North America, EU - Europe, AP - Asia Pacific, KR - South Korea, BR - Brazil, LAT - Latin America). Moreover, the listed values for `agent_1` and `gun_1` are valid playable [agents](https://playvalorant.com/en-us/agents/) and [weapons](https://playvalorant.com/en-us/arsenal/) according to Valorant's official website.

### Check for Incorrect Data Type
It is important to check for the data type of each variable since there could be possible representations of numerical values in the dataset as text or string, and vice versa.

In [26]:
val_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 84068 entries, 0 to 85677
Data columns (total 38 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   region            84068 non-null  category
 1   name              84068 non-null  object  
 2   tag               84068 non-null  object  
 3   rating            84068 non-null  category
 4   damage_round      84068 non-null  float64 
 5   headshots         84068 non-null  int64   
 6   headshot_percent  84068 non-null  float64 
 7   aces              84068 non-null  int64   
 8   clutches          84068 non-null  int64   
 9   flawless          84068 non-null  int64   
 10  first_bloods      84068 non-null  int64   
 11  kills             84068 non-null  int64   
 12  deaths            84068 non-null  int64   
 13  assists           84068 non-null  int64   
 14  kd_ratio          84068 non-null  float64 
 15  kills_round       84068 non-null  float64 
 16  most_kills        84068 non

Using this general information on the dataset's data types, let's generate a list of variables that should not be represented as numerical values.

In [27]:
val_df.select_dtypes(exclude=['int', 'float']).columns.tolist()

['region',
 'name',
 'tag',
 'rating',
 'agent_1',
 'agent_2',
 'agent_3',
 'gun1_name',
 'gun2_name',
 'gun3_name']

Basing from the variables' names and descriptions, none of these should be represented numerically. Therefore, they are in correct data type.

In [28]:
val_df.select_dtypes(include=['int', 'float']).columns.tolist()

['damage_round',
 'headshots',
 'headshot_percent',
 'aces',
 'clutches',
 'flawless',
 'first_bloods',
 'kills',
 'deaths',
 'assists',
 'kd_ratio',
 'kills_round',
 'most_kills',
 'score_round',
 'wins',
 'win_percent',
 'gun1_head',
 'gun1_body',
 'gun1_legs',
 'gun1_kills',
 'gun2_head',
 'gun2_body',
 'gun2_legs',
 'gun2_kills',
 'gun3_head',
 'gun3_body',
 'gun3_legs',
 'gun3_kills']

According to the variables' names and description, they should be represented numerically. Hence, they are in proper data type.

### Check for Missing Values
Let us check the size of the dataset before cleaning missing values.

In [9]:
val_df.shape

(85678, 38)

There are 85678 rows and 38 columns.

We can now begin checking for missing values.

In [10]:
# Return the raw rows with missing values
missing_df = val_df.isnull().sum()
missing_df[missing_df > 0]

name          2
tag         320
agent_2    1987
agent_3    5217
dtype: int64

There is something suspicious with the `tag` column. There are a lot of "missing values".

In [11]:
null_df = val_df[val_df['tag'].isnull()]
null_df

region                 name  tag      rating  damage_round  headshots  \
304       NA  Wilma Caukfit#fk no  NaN  Immortal 3         117.8        372   
1005      NA          ylang#Luv U  NaN  Immortal 1         136.9       1135   
1059      NA            pup#ARI G  NaN  Immortal 1         134.1        398   
1371      NA        jawniss#luv u  NaN  Immortal 1         195.6         56   
1581      NA       LFT Comfy#NA 1  NaN  Immortal 1         136.8        273   
...      ...                  ...  ...         ...           ...        ...   
83617    LAT   RVN FroxeKzz#EZ GG  NaN  Immortal 1         143.6        874   
83630    LAT        TriggeR#One T  NaN  Immortal 1         146.5        381   
84040    LAT         Maniex#on ig  NaN  Immortal 2         148.5        731   
84289    LAT           keii#on ig  NaN  Immortal 2         164.3        685   
84649    LAT  bife ツ#ツ 強力  NaN  Immortal 2         138.2        346   

       headshot_percent  aces  clutches  flawless  ...  gun2_name  gun2_head  \
304                25.6     0        63        26  ...     Vandal         38   
1005               24.2     3       135        92  ...     Vandal         40   
1059               21.6     1        78        43  ...    Phantom         31   
1371               26.9     0         4         3  ...   Operator          0   
1581               17.9     0        47        27  ...    Phantom         22   
...                 ...   ...       ...       ...  ...        ...        ...   
83617              21.4     1       110        72  ...     Vandal         30   
83630              15.6     1        65        35  ...     Vandal         26   
84040              24.9     0        91        61  ...    Phantom         28   
84289              25.1     6        74        68  ...    Phantom         27   
84649              18.3     1        52        35  ...    Phantom         25   

       gun2_body  gun2_legs  gun2_kills  gun3_name  gun3_head  gun3_body  \
304           59          3         161    Spectre         27         70   
1005          55          5         397    Spectre         22         71   
1059          66          3         161      Ghost         40         55   
1371         100          0          14    Spectre         14         86   
1581          73          4         108    Classic         26         71   
...          ...        ...         ...        ...        ...        ...   
83617         64          6         300    Spectre         24         71   
83630         68          5         199    Spectre         11         76   
84040         67          6         211    Spectre         26         68   
84289         68          5         132      Ghost         32         65   
84649         66          9          90    Spectre         16         69   

       gun3_legs  gun3_kills  
304            3          39  
1005           7          84  
1059           6          53  
1371           0           4  
1581           3          37  
...          ...         ...  
83617          5         148  
83630         13         122  
84040          6          86  
84289          3          58  
84649         15          47  

[320 rows x 38 columns]

It looks like the tag was appended to the name. We will fix this by splitting the `name` column into two columns.

In [12]:
# Split the name column into two columns
null_df = null_df['name'].str.split('#', expand=True)
null_df = null_df.rename(columns={0:'name', 1:'tag'})
null_df.loc[:, 'tag'] = '#' + null_df['tag']
null_df

name          tag
304    Wilma Caukfit       #fk no
1005           ylang       #Luv U
1059             pup       #ARI G
1371         jawniss       #luv u
1581       LFT Comfy        #NA 1
...              ...          ...
83617   RVN FroxeKzz       #EZ GG
83630        TriggeR       #One T
84040         Maniex       #on ig
84289           keii       #on ig
84649       bife ツ  #ツ 強力

[320 rows x 2 columns]

Now it looks like the `tag` column is fixed. We will replace the `tag` column in the original dataframe with the fixed `tag` column.

In [13]:
# Replace the name and tag column in the original dataframe with the fixed tag column
val_df.loc[val_df['tag'].isnull(), ['name', 'tag']] = null_df
val_df

region            name     tag      rating  damage_round  headshots  \
0         NA        ShimmyXD    #NA1     Radiant         135.8        992   
1         NA       XSET Cryo  #cells     Radiant         170.3        879   
2         NA   PuRelittleone  #yoruW     Radiant         147.5        720   
3         NA            Boba   #0068     Radiant         178.2        856   
4         NA     i love mina  #kelly     Radiant         149.8        534   
...      ...             ...     ...         ...           ...        ...   
85673    LAT        Kazutora  #img0d     Radiant         138.2        342   
85674    LAT  el lobo marino  #uthur     Radiant         182.9        650   
85675    LAT          p9pzet   #666x     Radiant         158.8        613   
85676    LAT  EZ4TGD EnSBuwu  #kmeve  Immortal 3         155.9        132   
85677    LAT            Neon    #SSJ     Radiant         164.4       1127   

       headshot_percent  aces  clutches  flawless  ...  gun2_name  gun2_head  \
0                  24.9     0       140        80  ...    Phantom         33   
1                  28.3     2       122        94  ...   Operator          8   
2                  24.0     3       117        59  ...    Phantom         36   
3                  37.3     3        83        49  ...    Sheriff         48   
4                  24.4     2        71        38  ...    Spectre         21   
...                 ...   ...       ...       ...  ...        ...        ...   
85673              21.4     0        58        54  ...     Vandal         28   
85674              30.1     4        77        42  ...     Vandal         40   
85675              30.2     0        70        54  ...    Phantom         40   
85676              22.2     0        23        18  ...    Phantom         39   
85677              20.7     3       136        82  ...   Operator          9   

       gun2_body  gun2_legs  gun2_kills  gun3_name  gun3_head  gun3_body  \
0             62          5         220    Classic         36         60   
1             91          0         226    Phantom         32         63   
2             61          3         231   Operator          8         91   
3             51          1          48    Phantom         44         56   
4             71          8          65   Operator          8         92   
...          ...        ...         ...        ...        ...        ...   
85673         69          2         175    Classic         39         59   
85674         57          3         212    Spectre         33         63   
85675         56          4         159   Operator         10         89   
85676         57          3          37    Spectre         27         69   
85677         88          2         132    Spectre         27         70   

       gun3_legs  gun3_kills  
0              3         147  
1              5         137  
2              1         102  
3              0          36  
4              0          64  
...          ...         ...  
85673          2          71  
85674          5         139  
85675          1          87  
85676          4          17  
85677          2         108  

[85678 rows x 38 columns]

Let's check again for missing values.

In [14]:
missing_df = val_df.isnull().sum()
missing_df[missing_df > 0]

name          2
tag           2
agent_2    1987
agent_3    5217
dtype: int64

There are still missing values in `name` and `tag`. Let's check the rows with missing values.

First, we will check the rows with missing values in `name`.

In [15]:
null_df = val_df[val_df['name'].isnull()]
null_df

region name  tag      rating  damage_round  headshots  headshot_percent  \
24453     EU  NaN  NaN  Immortal 1         152.2         17              22.5   
63603     AP  NaN  NaN      Gold 3         115.7        257              18.2   

       aces  clutches  flawless  ...  gun2_name  gun2_head  gun2_body  \
24453     0         4         1  ...      Ghost         60         20   
63603     0        52        28  ...    Classic         39         58   

       gun2_legs  gun2_kills  gun3_name  gun3_head  gun3_body  gun3_legs  \
24453         20           3    Spectre          8         85          8   
63603          3          47    Phantom         24         68          8   

       gun3_kills  
24453           3  
63603          39  

[2 rows x 38 columns]

There is at least one row with a missing `name` and `tag`.
We will fill the missing names with the value "(No Name)" and the missing tags with a string with format `#n` where n is the index of the row.

In [16]:
# Iterate through the rows and fill the missing names and tags
for i, row in null_df.iterrows():
    val_df.loc[i, 'name'] = f'(No Name)'
    val_df.loc[i, 'tag'] = f'#{i}'
val_df

region            name     tag      rating  damage_round  headshots  \
0         NA        ShimmyXD    #NA1     Radiant         135.8        992   
1         NA       XSET Cryo  #cells     Radiant         170.3        879   
2         NA   PuRelittleone  #yoruW     Radiant         147.5        720   
3         NA            Boba   #0068     Radiant         178.2        856   
4         NA     i love mina  #kelly     Radiant         149.8        534   
...      ...             ...     ...         ...           ...        ...   
85673    LAT        Kazutora  #img0d     Radiant         138.2        342   
85674    LAT  el lobo marino  #uthur     Radiant         182.9        650   
85675    LAT          p9pzet   #666x     Radiant         158.8        613   
85676    LAT  EZ4TGD EnSBuwu  #kmeve  Immortal 3         155.9        132   
85677    LAT            Neon    #SSJ     Radiant         164.4       1127   

       headshot_percent  aces  clutches  flawless  ...  gun2_name  gun2_head  \
0                  24.9     0       140        80  ...    Phantom         33   
1                  28.3     2       122        94  ...   Operator          8   
2                  24.0     3       117        59  ...    Phantom         36   
3                  37.3     3        83        49  ...    Sheriff         48   
4                  24.4     2        71        38  ...    Spectre         21   
...                 ...   ...       ...       ...  ...        ...        ...   
85673              21.4     0        58        54  ...     Vandal         28   
85674              30.1     4        77        42  ...     Vandal         40   
85675              30.2     0        70        54  ...    Phantom         40   
85676              22.2     0        23        18  ...    Phantom         39   
85677              20.7     3       136        82  ...   Operator          9   

       gun2_body  gun2_legs  gun2_kills  gun3_name  gun3_head  gun3_body  \
0             62          5         220    Classic         36         60   
1             91          0         226    Phantom         32         63   
2             61          3         231   Operator          8         91   
3             51          1          48    Phantom         44         56   
4             71          8          65   Operator          8         92   
...          ...        ...         ...        ...        ...        ...   
85673         69          2         175    Classic         39         59   
85674         57          3         212    Spectre         33         63   
85675         56          4         159   Operator         10         89   
85676         57          3          37    Spectre         27         69   
85677         88          2         132    Spectre         27         70   

       gun3_legs  gun3_kills  
0              3         147  
1              5         137  
2              1         102  
3              0          36  
4              0          64  
...          ...         ...  
85673          2          71  
85674          5         139  
85675          1          87  
85676          4          17  
85677          2         108  

[85678 rows x 38 columns]

Let's check again for missing values.

In [17]:
missing_df = val_df.isnull().sum()
missing_df[missing_df > 0]

agent_2    1987
agent_3    5217
dtype: int64

Looks like that's it! The current missing values are in the `agent_2` and `agent_3` columns.
These are acceptable because they may be players who only played one agent.

After cleaning the missing values, let's check the size of the dataset again.

In [18]:
val_df.shape

(85678, 38)

There are 85678 rows and 38 columns.

Here is the cleaned dataset.

In [19]:
val_df

region            name     tag      rating  damage_round  headshots  \
0         NA        ShimmyXD    #NA1     Radiant         135.8        992   
1         NA       XSET Cryo  #cells     Radiant         170.3        879   
2         NA   PuRelittleone  #yoruW     Radiant         147.5        720   
3         NA            Boba   #0068     Radiant         178.2        856   
4         NA     i love mina  #kelly     Radiant         149.8        534   
...      ...             ...     ...         ...           ...        ...   
85673    LAT        Kazutora  #img0d     Radiant         138.2        342   
85674    LAT  el lobo marino  #uthur     Radiant         182.9        650   
85675    LAT          p9pzet   #666x     Radiant         158.8        613   
85676    LAT  EZ4TGD EnSBuwu  #kmeve  Immortal 3         155.9        132   
85677    LAT            Neon    #SSJ     Radiant         164.4       1127   

       headshot_percent  aces  clutches  flawless  ...  gun2_name  gun2_head  \
0                  24.9     0       140        80  ...    Phantom         33   
1                  28.3     2       122        94  ...   Operator          8   
2                  24.0     3       117        59  ...    Phantom         36   
3                  37.3     3        83        49  ...    Sheriff         48   
4                  24.4     2        71        38  ...    Spectre         21   
...                 ...   ...       ...       ...  ...        ...        ...   
85673              21.4     0        58        54  ...     Vandal         28   
85674              30.1     4        77        42  ...     Vandal         40   
85675              30.2     0        70        54  ...    Phantom         40   
85676              22.2     0        23        18  ...    Phantom         39   
85677              20.7     3       136        82  ...   Operator          9   

       gun2_body  gun2_legs  gun2_kills  gun3_name  gun3_head  gun3_body  \
0             62          5         220    Classic         36         60   
1             91          0         226    Phantom         32         63   
2             61          3         231   Operator          8         91   
3             51          1          48    Phantom         44         56   
4             71          8          65   Operator          8         92   
...          ...        ...         ...        ...        ...        ...   
85673         69          2         175    Classic         39         59   
85674         57          3         212    Spectre         33         63   
85675         56          4         159   Operator         10         89   
85676         57          3          37    Spectre         27         69   
85677         88          2         132    Spectre         27         70   

       gun3_legs  gun3_kills  
0              3         147  
1              5         137  
2              1         102  
3              0          36  
4              0          64  
...          ...         ...  
85673          2          71  
85674          5         139  
85675          1          87  
85676          4          17  
85677          2         108  

[85678 rows x 38 columns]

### Remove Duplicate Rows
Let us check for duplicate rows.

In [20]:
val_df[val_df.duplicated()]

region              name    tag      rating  damage_round  headshots  \
327       NA            WRE Px  #1TAP  Immortal 3         133.8        368   
471       NA               mer  #goat  Immortal 2         148.9        295   
1828      NA           Misssso   #NA1  Immortal 1         149.9        512   
1839      NA         OnionBean   #NA1  Immortal 1         136.6        450   
2034      NA  OryxTheMadHatter   #NA1  Immortal 1         152.4       1120   
...      ...               ...    ...         ...           ...        ...   
83999    LAT              fraZ  #PENA  Immortal 1         157.2        395   
84062    LAT              crip  #2858  Immortal 2         140.7        180   
84064    LAT         Nicobass1   #LAN  Immortal 2         207.9        105   
84541    LAT           sshinko   #huh  Immortal 2         135.7        324   
84967    LAT        blablayost   #LAN  Immortal 3         150.2        383   

       headshot_percent  aces  clutches  flawless  ...  gun2_name  gun2_head  \
327                38.0     1        52        17  ...     Vandal         50   
471                26.6     0        29        20  ...    Phantom         40   
1828               17.3     1        75        48  ...     Vandal         29   
1839               19.3     1        56        36  ...    Classic         28   
2034               21.5     2       151        88  ...    Phantom         30   
...                 ...   ...       ...       ...  ...        ...        ...   
83999              25.8     0        51        20  ...    Phantom         31   
84062              28.5     0        21        14  ...    Phantom         34   
84064              18.2     0        15         4  ...    Spectre         12   
84541              20.5     2        58        33  ...    Phantom         28   
84967              27.3     0        43        40  ...    Phantom         41   

       gun2_body  gun2_legs  gun2_kills  gun3_name  gun3_head  gun3_body  \
327           48          3         141    Spectre         39         58   
471           57          3          90    Spectre         29         68   
1828          66          4         260    Spectre         19         73   
1839          67          5          79    Spectre         19         72   
2034          64          6         388    Spectre         20         70   
...          ...        ...         ...        ...        ...        ...   
83999         66          3          60      Ghost         37         59   
84062         64          3          68      Ghost         29         66   
84064         78         10          25      Ghost         63         31   
84541         66          6         161    Spectre         17         77   
84967         56          2          73   Operator          6         94   

       gun3_legs  gun3_kills  
327            4          55  
471            3          44  
1828           8          77  
1839           9          56  
2034          10         137  
...          ...         ...  
83999          4          45  
84062          6          15  
84064          6          10  
84541          6          53  
84967          0          50  

[1084 rows x 38 columns]

We will remove duplicate rows based on the `name` and `tag` columns. These columns are used to identify a player.
Then, we will keep the first row of the duplicate rows.

In [21]:
# Use loc indexer to avoid SettingWithCopyWarning
val_df = val_df.loc[~val_df.duplicated(subset=['name', 'tag'], keep='first')]
val_df

region            name     tag      rating  damage_round  headshots  \
0         NA        ShimmyXD    #NA1     Radiant         135.8        992   
1         NA       XSET Cryo  #cells     Radiant         170.3        879   
2         NA   PuRelittleone  #yoruW     Radiant         147.5        720   
3         NA            Boba   #0068     Radiant         178.2        856   
4         NA     i love mina  #kelly     Radiant         149.8        534   
...      ...             ...     ...         ...           ...        ...   
85673    LAT        Kazutora  #img0d     Radiant         138.2        342   
85674    LAT  el lobo marino  #uthur     Radiant         182.9        650   
85675    LAT          p9pzet   #666x     Radiant         158.8        613   
85676    LAT  EZ4TGD EnSBuwu  #kmeve  Immortal 3         155.9        132   
85677    LAT            Neon    #SSJ     Radiant         164.4       1127   

       headshot_percent  aces  clutches  flawless  ...  gun2_name  gun2_head  \
0                  24.9     0       140        80  ...    Phantom         33   
1                  28.3     2       122        94  ...   Operator          8   
2                  24.0     3       117        59  ...    Phantom         36   
3                  37.3     3        83        49  ...    Sheriff         48   
4                  24.4     2        71        38  ...    Spectre         21   
...                 ...   ...       ...       ...  ...        ...        ...   
85673              21.4     0        58        54  ...     Vandal         28   
85674              30.1     4        77        42  ...     Vandal         40   
85675              30.2     0        70        54  ...    Phantom         40   
85676              22.2     0        23        18  ...    Phantom         39   
85677              20.7     3       136        82  ...   Operator          9   

       gun2_body  gun2_legs  gun2_kills  gun3_name  gun3_head  gun3_body  \
0             62          5         220    Classic         36         60   
1             91          0         226    Phantom         32         63   
2             61          3         231   Operator          8         91   
3             51          1          48    Phantom         44         56   
4             71          8          65   Operator          8         92   
...          ...        ...         ...        ...        ...        ...   
85673         69          2         175    Classic         39         59   
85674         57          3         212    Spectre         33         63   
85675         56          4         159   Operator         10         89   
85676         57          3          37    Spectre         27         69   
85677         88          2         132    Spectre         27         70   

       gun3_legs  gun3_kills  
0              3         147  
1              5         137  
2              1         102  
3              0          36  
4              0          64  
...          ...         ...  
85673          2          71  
85674          5         139  
85675          1          87  
85676          4          17  
85677          2         108  

[84068 rows x 38 columns]

Let's check for duplicate rows again.

In [22]:
val_df[val_df.duplicated()]

Empty DataFrame
Columns: [region, name, tag, rating, damage_round, headshots, headshot_percent, aces, clutches, flawless, first_bloods, kills, deaths, assists, kd_ratio, kills_round, most_kills, score_round, wins, win_percent, agent_1, agent_2, agent_3, gun1_name, gun1_head, gun1_body, gun1_legs, gun1_kills, gun2_name, gun2_head, gun2_body, gun2_legs, gun2_kills, gun3_name, gun3_head, gun3_body, gun3_legs, gun3_kills]
Index: []

[0 rows x 38 columns]

There are no more duplicate rows.

After removing duplicate rows, let's check the size of the dataset again.

In [23]:
val_df.shape

(84068, 38)

In [24]:
def check_nan_or_non_integer(column):
    numeric_column = pd.to_numeric(column, errors='coerce')
    has_non_integer_values = numeric_column.isna().any()
    return has_non_integer_values

columns_to_check = ['damage_round', 'headshots', 'headshot_percent', 'aces', 'clutches', 'flawless', 'first_bloods', 'kills', 'deaths', 'assists', 'kd_ratio', 'kills_round', 'most_kills', 'kills', 'score_round', 'wins', 'gun1_head', 'gun1_body', 'gun1_legs', 'gun1_kills', 'gun2_head', 'gun2_body', 'gun2_legs', 'gun2_kills', 'gun3_head', 'gun3_body', 'gun3_legs', 'gun3_kills']

for column in columns_to_check:
    result = check_nan_or_non_integer(val_df[column])
    print("Has non integer values " + f"{column}: {result}")


Has non integer values damage_round: False
Has non integer values headshots: False
Has non integer values headshot_percent: False
Has non integer values aces: False
Has non integer values clutches: False
Has non integer values flawless: False
Has non integer values first_bloods: False
Has non integer values kills: False
Has non integer values deaths: False
Has non integer values assists: False
Has non integer values kd_ratio: False
Has non integer values kills_round: False
Has non integer values most_kills: False
Has non integer values kills: False
Has non integer values score_round: False
Has non integer values wins: False
Has non integer values gun1_head: False
Has non integer values gun1_body: False
Has non integer values gun1_legs: False
Has non integer values gun1_kills: False
Has non integer values gun2_head: False
Has non integer values gun2_body: False
Has non integer values gun2_legs: False
Has non integer values gun2_kills: False
Has non integer values gun3_head: False
Has no

In [25]:
print(val_df['rating'].unique().tolist())
print(val_df['agent_1'].unique().tolist())
print(val_df['agent_2'].unique().tolist())
print(val_df['agent_3'].unique().tolist())
print(val_df['gun1_name'].unique().tolist())
print(val_df['gun2_name'].unique().tolist())
print(val_df['gun3_name'].unique().tolist())


['Radiant', 'Immortal 3', 'Immortal 2', 'Immortal 1', 'Silver 2', 'Gold 2', 'Diamond 3', 'Gold 1', 'Unrated', 'Platinum 1', 'Bronze 3', 'Gold 3', 'Bronze 2', 'Diamond 2', 'Silver 3', 'Silver 1', 'Platinum 2', 'Diamond 1', 'Platinum 3']
['Fade', 'Chamber', 'Yoru', 'Jett', 'Sage', 'KAY/O', 'Sova', 'Raze', 'Omen', 'Breach', 'Reyna', 'Neon', 'Skye', 'Viper', 'Brimstone', 'Phoenix', 'Astra', 'Killjoy', 'Cypher']
['Viper', 'Jett', 'Chamber', 'Raze', 'Sova', 'Fade', 'Reyna', 'Skye', 'Omen', 'Neon', 'KAY/O', 'Brimstone', nan, 'Sage', 'Cypher', 'Astra', 'Killjoy', 'Breach', 'Yoru', 'Phoenix']
['Omen', 'Raze', 'Chamber', 'KAY/O', 'Sova', 'Sage', 'Jett', 'Reyna', 'Viper', 'Neon', 'Fade', nan, 'Skye', 'Brimstone', 'Breach', 'Killjoy', 'Cypher', 'Astra', 'Yoru', 'Phoenix']
['Vandal', 'Operator', 'Phantom', 'Judge', 'Odin', 'Guardian', 'Spectre', 'Classic', 'Ghost', 'Sheriff', 'Marshal', 'Shorty', 'Bulldog', 'Ares', 'Frenzy', 'Bucky']
['Phantom', 'Operator', 'Sheriff', 'Spectre', 'Vandal', 'Classic'